### PR TITLE
Fix `WITH DOCKER` loading when image contains port number

### PR DESCRIPTION
**Pull request by vladaionescu**
Fixes https://github.com/earthly/earthly/issues/2071

This PR goes the extra mile to maintain forwards-backwards compatibility given the usage of the feature flag via satellites. Besides the automated test added, I have also manually tested old earthly + new buildkit and new earthly + old buildkit (this particular bug is not fixed in these combinations, but everything else still works).

Minor hidden addition as part of this PR: a way to inhibit buildkit updating and restarting (`earthly --no-buildkit-update`) - this helps test an old buildkit with a new earthly.

### CODE REVIEW

File: builder/image_solver.go [hunk 1]
```diff
@@ -198,24 +198,36 @@ func (m *multiImageSolver) SolveImages(ctx context.Context, imageDefs []*states.
...
+	} else {
+		switch info.BuildkitVersion.Version {
+		case "v0.6.19":
+		case "v0.6.20":
+		default:
+			newInterImgFormat = true
+		}
+	}
+
+	onPullMap := make(map[string]string) // intermediate image name -> final image name
```
Comment (alexcb) [inline builder/image_solver.go:214]: is this duplicating https://github.com/earthly/earthly/blob/main/util/gatewaycrafter/pullpingmap.go ?

we have a similar onPull function under builder, where we pull images out of the pullpingmap: https://github.com/earthly/earthly/blob/main/builder/builder.go#L412-L416

Comment (vladaionescu) [reply builder/image_solver.go:214]: It is very similar but different enough. The intention is the same - to keep track of a mapping between intermediate names vs final (re-tagged) names. However, the images are not exported locally - they are exported into `WITH DOCKER` - so the code you're referencing couldn't be reused in this situation.

**Comment by vladaionescu** [comment]
Note that the satellite failure is expected. This is what this PR is fixing (after a satellite update with the fix in this PR, this failure would not show up anymore).

Screen Shot 2022-08-03 at 4 21 25 PM

File: buildkitd/dockerd-wrapper.sh [hunk 2]
```diff
@@ -193,8 +193,18 @@ load_registry_images() {
     if [ -n "$EARTHLY_DOCKER_LOAD_REGISTRY" ]; then
         echo "Loading images from BuildKit via embedded registry..."
         for img in $EARTHLY_DOCKER_LOAD_REGISTRY; do
-            user_tag=$(printf '%s' "$img" | cut -d'/' -f2-)
-            with_reg="$buildkit_docker_registry/$img"
+            case "$img" in
+                *'|'*)
+                    with_reg="$buildkit_docker_registry"/$(printf '%s' "$img" | cut -d'|' -f1)
+                    user_tag=$(printf '%s' "$img" | cut -d'|' -f2-)
```
Comment (alexcb) [inline buildkitd/dockerd-wrapper.sh:199]:
```suggestion
                    user_tag="$(printf '%s' "$img" | cut -d'|' -f2-)"
```

File: buildkitd/dockerd-wrapper.sh [hunk 3]
```diff
@@ -193,8 +193,18 @@ load_registry_images() {
...
+                    ;;
+                *)
+                    # Old format before v0.6.21.
+                    with_reg="$buildkit_docker_registry/$img"
+                    user_tag=$(printf '%s' "$img" | cut -d'/' -f2-)
```
Comment (alexcb) [inline buildkitd/dockerd-wrapper.sh:204]:
```suggestion
                    user_tag="$(printf '%s' "$img" | cut -d'/' -f2-)"
```

Comment (vladaionescu) [reply buildkitd/dockerd-wrapper.sh:204]: I think it doesn't work due to the `"$img"` in the middle - not what's the correct way to quote this (or if it even needs quotes in this case)

Comment (alexcb) [reply buildkitd/dockerd-wrapper.sh:204]: that surprises me, we have similar bits of shell that use double quotes both around and inside a shell out: https://github.com/earthly/earthly/blob/main/tests/secret-provider-config/test-secret-provider#L25

Comment (alexcb) [reply buildkitd/dockerd-wrapper.sh:204]: anyways, if it's passing our linter and tests work might be good enough.

**Review by alexcb — APPROVED** [review]

File: buildkitd/dockerd-wrapper.sh (hunk 3 above)
Comment (vladaionescu) [reply buildkitd/dockerd-wrapper.sh]: TIL!

Fixed

**Pull request merged**